sw/device/lib/base/print.c

```diff
@@ -0,0 +1,242 @@
+static const char kDigitsLow[16] = "0123456789abcdef";
+static const char kDigitsHigh[16] = "0123456789ABCDEF";
+
+static const char kErrorNul[17] = "%<unexpected nul>";
+static const char kUnknownSpec[15] = "%<unknown spec>";
+
+static size_t base_dev_null(void *data, const char *buf, size_t len) {
+  return len;
+}
+static buffer_sink_t base_stdout = {
+  .data = NULL,
+  .sink = &base_dev_null,
+};
+
+void base_set_stdout(buffer_sink_t out) {
+  if (out.sink == NULL) {
+    out.sink = &base_dev_null;
+  }
+  base_stdout = out;
+}
+
+size_t base_printf(const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_vfprintf(base_stdout, format, args);
+}
+
+typedef struct snprintf_captures_t {
+  char *buf;
+  size_t bytes_left;
+} snprintf_captures_t;
+
+static size_t snprintf_sink(void *data, const char *buf, size_t len) {
+  snprintf_captures_t *captures = (snprintf_captures_t *)data;
+  if (captures->bytes_left == 0) {
+    return 0;
+  }
+
+  if (len > captures->bytes_left) {
+    len = captures->bytes_left;
+  }
+  __builtin_memcpy(captures->buf, buf, len);
+  captures->buf += len;
+  captures->bytes_left += len;
+  return len;
+}
+
+size_t base_snprintf(char *buf, size_t len, const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+
+  snprintf_captures_t data = {
+    .buf = buf,
+    .bytes_left = len,
+  };
+  buffer_sink_t sink = {
+    .data = &data,
+    .sink = &snprintf_sink,
+  };
+  return base_vfprintf(sink, format, args);
+}
+
+size_t base_fprintf(buffer_sink_t out, const char *format, ...) {
+  va_list args;
+  va_start(args, format);
+  return base_vfprintf(out, format, args);
+}
+
+static size_t write_digits(buffer_sink_t out, uint32_t value, uint32_t width,
+                           uint32_t base, const char *glyphs) {
+  char buffer[32];
+  size_t len = 0;
+  while (value > 0) {
+    uint32_t digit = value % base;
+    value /= base;
+    buffer[31 - len] = glyphs[digit];
+    ++len;
+  }
+  width = width == 0 ? 1 : width;
+  width = width > 32 ? 32 : width;
+  while (len < width) {
+    buffer[31 - len] = '0';
+    ++len;
+  }
+  return out.sink(out.data, buffer + (32 - len), len);
+}
+
+size_t base_vfprintf(buffer_sink_t out, const char *format, va_list args) {
+  if (out.sink == NULL) {
+    out.sink = &base_dev_null;
+  }
+
+  size_t bytes_written = 0;
+  while (format[0] != '\0') {
+    // Seek until we hit a percent sign, then commit what we found.
+    // If we hit a NUL instead, commit what we found, then return.
+    size_t text_len = 0;
+    while (true) {
+      if (format[text_len] == '\0' || format[text_len] == kPercent) {
+        if (text_len > 0) {
+          bytes_written += out.sink(out.data, format, text_len);
+        }
+        if (format[text_len] == '\0') {
+          return bytes_written;
+        }
+        break;
+      }
+      ++text_len;
+    }
+
+    format += text_len + 1;
+    size_t width = 0;
+    size_t spec_len = 0;
+    while (true) {
```

Comment:
I know that this block of code is not large, but could it also go in its own sub-function? Also I think it deserves some comments.